allennlp/nlp_api.py

```diff
@@ -0,0 +1,289 @@
+from typing import Callable, Dict, List, Union
+
+import torch
+
+from .common.checks import ConfigurationError
+from .common import Params
+
+
+class NlpApi:
+    """
+    The ``NlpApi`` provides a high-level API for building NLP models.  You don't have to use it
+    for building a model with AllenNLP, but we think you should, because it's useful.
+
+    ``NlpApi`` abstracts away particular decisions that are frequently necessary in NLP, like how
```

Comment:
Obviously it's hard to comment on the overall design if there is no concrete example to illustrate the API usage [and from_params is not implemented]. I'd prioritize writing and informally reviewing a few API straw-clients before committing to API details.

I kind of get what you're saying and it sounds nice, the "name" is a pointer in a params thing [maybe?]. Then I look at the APIs downstream and I see:
```
def embed_text_field(self, input_tensors, name, fallback_behavior)
def get_recurrent_layer(self, name, fallback_behavior)
def get_recurrent_cell(self,   name, fallback_behavior)
def from_params(cls, params: Params)
```
If there are concrete objects passed somewhere in this API, they certainly are hiding really well ;)